evm_arithmetization/src/prover.rs

```diff
@@ -522,6 +523,92 @@ fn build_segment_data<F: RichField>(
     }
 }
 
+pub struct SegmentDataIterator {
+    pub partial_next_data: Option<GenerationSegmentData>,
+    pub inputs: GenerationInputs,
```

Comment:
Do we really need this? This forces a clone on every iteration, kind of wasteful.

Reply:
Yes, we need to maintain the full interpreter state to be able to keep running it without backtracking (notably for the prover inputs or the jumpdest table). We need a clone anyway every time the iterator returns an item, meaning that you only have one extra clone per txn.

Reply:
This is not really what I'm asking. The interpreter takes it as a reference already. Why not storing a reference to the inputs as well. We don't need the iterator to return the `GenerationInputs`, we know which ones they are for a given payload and can retrieve them on the `zero-bin` side.

Reply:
The reason why it's in the iterator is actually for zero-bin: the Iterator is turned into an `IndexedStream` which should return `(GenerationInputs, GenerationSegmentData)`, since this is what `SegmentProof` takes as `Input`.

I tried something like this on zero-bin (with `SegmentDataIterator` taking `&GenerationInputs` instead of `GenerationInputs`):
`  let tx_proof_futs: FuturesUnordered<_> = txs
            .into_iter()
            .enumerate()
            .map(|(idx, txn)| {
                let data_iterator = SegmentDataIterator {
                    partial_next_data: None,
                    inputs: &txn,
                    max_cpu_len_log: Some(max_cpu_len_log),
                };

                Directive::map(IndexedStream::from(data_iterator), &SegmentProof)
                    .fold(&ops::SegmentAggProof)
                    .run(runtime)
                    .map(move |e| {
                        e.map(|p| (idx, proof_gen::proof_types::TxnAggregatableProof::from(p)))
                    })
            })
            .collect();`
But I get: `cannot return value referencing temporary value` (the referenced value being `&txn`). So I am not sure how to get rid of the clones / how to fix this error

Reply:
Can't we define the `inputs` as a field of `SegmentProof` (same as `save_inputs_on_error`), so that we save all these clones?

Reply:
I also tried that, and I got the same error :(

Reply:
Kind of a skim here and I can't test the code, but would `iter()` instead of `into_iter()` work here? Then the reference `txn` doesn't go out of scope after each iteration.

Otherwise maybe we can put `txn` in an `Arc<_>` if we're desperate? :sweat_smile: